File: src/python/problems/h_Index_test.py
# https://leetcode.com/problems/h-index

from core.problem_base import *

class HIndex(ProblemBase):
    def Solution(self, citations: list[int]) -> int:
        citations.sort()
        length = len(citations)

        for i in range(length):
            if citations[i] < length - i:
                continue

            for hIndex in range(citations[i], -1, -1):
                if hIndex <= length - i:
                    return hIndex

        return 0

if __name__ == '__main__':
    TestGen(HIndex) \
        .Add(lambda tc: tc.Param([3, 1, 7, 8, 9]).Result(3)) \
        .Add(lambda tc: tc.Param([3,0,6,1,5]).Result(3)) \
        .Add(lambda tc: tc.Param([1,3,1]).Result(1)) \
        .Add(lambda tc: tc.Param([100]).Result(1)) \
        .Run()

File: src/go/challenges/problems/h_index_test.go
// https://leetcode.com/problems/h-index

package problems

import (
	"sort"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type HIndex struct{}

func TestHIndex(t *testing.T) {
	gen := core.TestSuite[HIndex]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{3, 0, 6, 1, 5}).Result(3)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 3, 1}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{100}).Result(1)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{3, 1, 7, 8, 9}).Result(3)
	}).Run(t)
}

func (HIndex) Solution(citations []int) int {
	length := len(citations)

	sort.Ints(citations)

	for i := 0; i < length; i++ {
		if citations[i] < length-i {
			continue
		}

		for hIndex := citations[i]; hIndex >= 0; hIndex-- {
			if hIndex <= length-i {
				return hIndex
			}
		}
	}

	return 0
}
